File: codeX-frontend/src/services/native/git.js
/**
 * Native Git Service
 * Manages local Git operations directly via local Git binary on desktop.
 * Never transmits Git credentials or repos across remote collaboration websockets.
 */

import { isDesktopApp } from './platform';

export const gitService = {
  isSupported() {
    return isDesktopApp();
  },

  async getStatus(repoPath) {
    if (!isDesktopApp() || !repoPath) {
      return {
        is_repo: false,
        branch: '',
        ahead: 0,
        behind: 0,
        staged: [],
        unstaged: [],
        untracked: [],
      };
    }
    const { invoke } = await import('@tauri-apps/api/core');
    return await invoke('git_status', { repoPath });
  },

  async getBranches(repoPath) {
    if (!isDesktopApp() || !repoPath) return [];
    const { invoke } = await import('@tauri-apps/api/core');
    return await invoke('git_branches', { repoPath });
  },

  async checkout(repoPath, branch) {
    if (!isDesktopApp() || !repoPath) {
      throw new Error('Git operations are only available on desktop');
    }
    const { invoke } = await import('@tauri-apps/api/core');
    return await invoke('git_checkout', { repoPath, branch });
  },

  async stage(repoPath, paths) {
    if (!isDesktopApp() || !repoPath) return;
    const { invoke } = await import('@tauri-apps/api/core');
    return await invoke('git_stage', { repoPath, paths });
  },

  async unstage(repoPath, paths) {
    if (!isDesktopApp() || !repoPath) return;
    const { invoke } = await import('@tauri-apps/api/core');
    return await invoke('git_unstage', { repoPath, paths });
  },

  async commit(repoPath, message) {
    if (!isDesktopApp() || !repoPath) {
      throw new Error('Git operations are only available on desktop');
    }
    const { invoke } = await import('@tauri-apps/api/core');
    return await invoke('git_commit', { repoPath, message });
  },

  async diff(repoPath, filePath = null, staged = false) {
    if (!isDesktopApp() || !repoPath) return '';
    const { invoke } = await import('@tauri-apps/api/core');
    return await invoke('git_diff', { repoPath, filePath, staged });
  },

  async log(repoPath, maxCount = 20) {
    if (!isDesktopApp() || !repoPath) return [];
    const { invoke } = await import('@tauri-apps/api/core');
    return await invoke('git_log', { repoPath, maxCount });
  },

  async fetch(repoPath) {
    if (!isDesktopApp() || !repoPath) return '';
    const { invoke } = await import('@tauri-apps/api/core');
    return await invoke('git_fetch', { repoPath });
  },

  async pull(repoPath) {
    if (!isDesktopApp() || !repoPath) return '';
    const { invoke } = await import('@tauri-apps/api/core');
    return await invoke('git_pull', { repoPath });
  },

  async push(repoPath) {
    if (!isDesktopApp() || !repoPath) return '';
    const { invoke } = await import('@tauri-apps/api/core');
    return await invoke('git_push', { repoPath });
  },
};

File: codeX-frontend/src-tauri/src/commands/git.rs
use crate::git::{
    self, GitCommitInfo, GitStatusResult,
};

#[tauri::command]
pub fn git_status(repo_path: String) -> Result<GitStatusResult, String> {
    git::get_status(&repo_path)
}

#[tauri::command]
pub fn git_branches(repo_path: String) -> Result<Vec<String>, String> {
    git::get_branches(&repo_path)
}

#[tauri::command]
pub fn git_checkout(repo_path: String, branch: String) -> Result<String, String> {
    git::checkout(&repo_path, &branch)
}

#[tauri::command]
pub fn git_stage(repo_path: String, paths: Vec<String>) -> Result<(), String> {
    git::stage(&repo_path, paths)
}

#[tauri::command]
pub fn git_unstage(repo_path: String, paths: Vec<String>) -> Result<(), String> {
    git::unstage(&repo_path, paths)
}

#[tauri::command]
pub fn git_commit(repo_path: String, message: String) -> Result<String, String> {
    git::commit(&repo_path, &message)
}

#[tauri::command]
pub fn git_diff(
    repo_path: String,
    file_path: Option<String>,
    staged: Option<bool>,
) -> Result<String, String> {
    git::diff(&repo_path, file_path, staged.unwrap_or(false))
}

#[tauri::command]
pub fn git_log(repo_path: String, max_count: Option<usize>) -> Result<Vec<GitCommitInfo>, String> {
    git::log(&repo_path, max_count.unwrap_or(20))
}

#[tauri::command]
pub fn git_fetch(repo_path: String) -> Result<String, String> {
    git::fetch(&repo_path)
}

#[tauri::command]
pub fn git_pull(repo_path: String) -> Result<String, String> {
    git::pull(&repo_path)
}

#[tauri::command]
pub fn git_push(repo_path: String) -> Result<String, String> {
    git::push(&repo_path)
}
